### ISSUE TITLE
[Bug] - Python venv fails on 2023.6.20241209.0

### ISSUE THREAD

**Issue by nhinds** [amazonlinux/amazon-linux-2023#861]
**Describe the bug**
Creating Python virtual environments with `python3 -m venv ...` fails on Amazon Linux 2023.6.20241209.0, while it succeeded on Amazon Linux 2023.6.20241121.0

**To Reproduce**
1. Run `python3 -m venv /tmp/.venv`

**Expected behavior**
A Python virtual environment is created

**Screenshots**
Fails on 2023.6.20241209.0:
![image](https://github.com/user-attachments/assets/fe1d8ecb-8fa0-45ea-991e-0461d9833262)

Succeeds on older 2023.6.20241121.0:
![image](https://github.com/user-attachments/assets/d5a7eeb6-3fdb-46cb-9b04-24660e2aa0ce)

**Desktop (please complete the following information):**
 - OS: Amazon Linux 2023.6.20241209.0
 - Browser N/A
 - Version 2023.6.20241209.0

**Additional context**
I originally experienced this in EC2 Image Builder while launching `arn:aws:imagebuilder:ap-southeast-2:aws:image/amazon-linux-2023-x86/x.x.x` (currently `ami-044b60a0a1b098489` "al2023-ami-2023.6.20241209.0-kernel-6.1-x86_64"), but this is also reproducible on the Docker image `public.ecr.aws/amazonlinux/amazonlinux:2023.6.20241209.0`. This can also be reproduced on EC2 instances launched from older AMIs that have run something like `dnf update --releasever=latest`.

The output on failure is:
```
bash-5.2# python3 -m venv /tmp/.venv
Error: Command '['/tmp/.venv/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
```
Running that command manually shows the following circular dependency error:
```
bash-5.2# /tmp/.venv/bin/python3 -Im ensurepip --upgrade --default-pip
/tmp/.venv/bin/python3: Error while finding module specification for 'ensurepip.__main__' (ImportError: cannot import name '_bundled' from partially initialized module 'ensurepip' (most likely due to a circular import) (/usr/lib64/python3.9/ensurepip/__init__.py))
```

Running the same commands on `public.ecr.aws/amazonlinux/amazonlinux:2023.6.20241121.0` works:
```
bash-5.2# python3 -m venv /tmp/.venv
bash-5.2# /tmp/.venv/bin/python3 -Im ensurepip --upgrade --default-pip
Looking in links: /tmp/tmpy5ff95ld
Requirement already satisfied: setuptools in /tmp/.venv/lib/python3.9/site-packages (59.6.0)
Requirement already satisfied: pip in /tmp/.venv/lib/python3.9/site-packages (21.3.1)
```

**Comment by nhinds** [comment]
Looking at the differences between `python3.9-3.9.16-1.amzn2023.0.9.src.rpm` and `python3.9-3.9.20-1.amzn2023.0.1.src.rpm`, it seems `00189-use-rpm-wheels.patch` is no longer being applied:
```
--- ../3.9.16/python3.9.spec	2024-07-30 03:47:40
+++ ../3.9.20/python3.9.spec	2024-12-05 09:10:06
...
@@ -812,14 +800,6 @@
 # Apply patches up to 188
 %autopatch -M 188

-%if %{with rpmwheels}
-# Temporary workaround for https://bugzilla.redhat.com/show_bug.cgi?id=1954999
-%{?!apply_patch:%define apply_patch(qp:m:) {%__apply_patch %**}}
-
-%apply_patch -q %{PATCH189}
-rm Lib/ensurepip/_bundled/*.whl
-%endif
-
 # Apply the remaining patches
 %autopatch -m 190
...
```

Among other changes to `ensurepip`, patch 189 would remove the `from . import _bundled` line from `Lib/ensurepip/__init__.py` which is failing on the newer version of Python:
```
...
--- a/Lib/ensurepip/__init__.py
+++ b/Lib/ensurepip/__init__.py
...
@@ -6,13 +8,29 @@ import tempfile
 import subprocess
 from importlib import resources

-from . import _bundled
-
...
```

**Comment by jplanckeel-ep** [comment]
We have the same problem

**Comment by elsaco** [comment]
This looks like a bad build of `python3.9.20` packages. I'm trying to build `python3.9-3.9.20-1.amzn2023.0.1.src.rpm` but there are  several issues with the `python3.9.spec`

The funny one is `bogus date in %changelog: Fri Nov 26 2024 Yash Gupta <ydgupta@amazon.com> - 3.9.20-1.amzn2023.0.1`

Nov 26, 2024, was a Tuesday!

**Comment by elsaco** [comment]
@nhinds the `python3.9-3.9.20-1.amzn2023.0.1.src.rpm` is based on `python3.9-3.9.20-1.fc39.src.rpm`. Looking at `python3.9.spec` there is a conditional build that will include patch 189, see line 794:

```
794 %if %{with rpmwheels}
795 %autopatch 189
796 rm Lib/ensurepip/_bundled/*.whl
797 %endif
```

The `amzn` spec file is missing this conditional. I don't know how the `amzn` package was build, with or without `rpmwheels`, but that seems to isolate the issue.

**Comment by nhinds** [comment]
Based on the other conditionals in `python3.9.spec`, it seems like it is being built with `rpmwheels` since `python3-libs` requires `python3-pip-wheel`:
```
bash-5.2# rpm -q --requires  python3-libs|grep pip
python3-pip-wheel >= 21.2.3-4
```
```
%package -n %{pkgname}-libs
Summary:        Python runtime libraries

%if %{with rpmwheels}
Requires: %{python_wheel_pkg_prefix}-setuptools-wheel >= 53.0.0-7
Requires: %{python_wheel_pkg_prefix}-pip-wheel >= 21.2.3-4
%else
Provides: bundled(python3dist(pip)) = %{pip_version}
Provides: bundled(python3dist(setuptools)) = %{setuptools_version}
%endif
```

Will need to wait for someone from Amazon to look at this - perhaps they could apply this patch from Fedora which replaced the `apply_patch` workaround with the call to `%autopatch 189`: https://src.fedoraproject.org/rpms/python3.9/c/7636db24d127095f4b1f6c71d8ca9e81e46aee99, rather than deleting the whole block

**Comment by elsaco** [comment]
While using `dnf update --releasever=2023.6.20241209.0` did pull some news packages, including the broken `python3-libs`, this release was never advertised by `dnf check-update`. 

There is a new release as a few hours ago, `20241212`:

```
WARNING:
  A newer release of "Amazon Linux" is available.

  Available Versions:

  Version 2023.6.20241212:
    Run the following command to upgrade to 2023.6.20241212:

      dnf upgrade --releasever=2023.6.20241212

    Release notes:
     https://docs.aws.amazon.com/linux/al2023/release-notes/relnotes-2023.6.20241212.html
```

Time to check  if `ensurepip` module issue was fixed!

**Comment by nhinds** [comment]
Good catch - yes, version `3.9.20-1.amzn2023.0.2` of the Python packages seems to have fixed this. Also mentioned in the changelog:
```
Changelogs for python3-3.9.20-1.amzn2023.0.2.aarch64
* Wed Dec 11 00:00:00 2024 Benjamin Herrenschmidt <benh@amazon.com> - 3.9.20-1.amzn2023.0.2
- Restore lost changes breaking pip,venv and FIPS support etc...
- Adjust fix for CVE-2023-27043 to make it disable-able via config or env

* Tue Nov 26 00:00:00 2024 Yash Gupta <ydgupta@amazon.com> - 3.9.20-1.amzn2023.0.1
- Bump to 3.9.20
```

**Comment by sktomer** [comment]
Amazon Linux is aware of this issue and actively working to mitigate.
We have rolled back the bad update and will soon follow up with the fixes.

**Comment by sktomer** [comment]
As of this moment, `public.ecr.aws/amazonlinux/amazonlinux:latest` points to `public.ecr.aws/amazonlinux/amazonlinux:2023.6.20241121.0`.

This is the version we recommend right now, we will update here once we have a validated new version with the latest security updates.

**Comment by naxrohan** [comment]
Got the same issue today.
`Stderr: Error: Command '['/tmp/ec2_utility/.venv/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
/tmp/AWSTxx/script-24xxx5.sh: line 11: .venv/bin/activate: No such file or directory
sudo: .venv/bin/pip3: command not found
`

**Comment by ozbenh** [comment]
Fix should be out now

**Comment by stewartsmith** [comment]
Fixed in the 2023.6.20241212 release
